### ISSUE TITLE
Culture-Specific String Comparisons Cause RollingFileAppender Failure in log4net on NET 5+ (Thai Culture)

### ISSUE THREAD

**Issue by stianeklund** [apache/logging-log4net#179]
Issue: log4net contains multiple instances of culture-specific string comparisons, which can cause errors when running in certain cultural settings. 

This issue particularly affects systems with the Thai (th-TH) culture on Windows. I created an issue in Jira last year: https://issues.apache.org/jira/browse/LOG4NET-694a but haven't seen any updates to it, and I wasn't sure whether or not it was being maintained anymore. Now seeing updates to the project on github I decided to create an issue here as well.

Affected Versions: This issue has been identified in log4net version `2.0.15`, but likely is also present in newer versions.

Details: 

The problem occurs for users running log4net on .NET 5 (and possibly .NET Core 3.0 and above). 
NET 5 introduced a change in the way the runtime handles culture by switching from NLS (*National Language Support*) to ICU (*International Components for Unicode*). More information on this transition can be found here: [ICU in .NET](https://learn.microsoft.com/en-us/dotnet/core/extensions/globalization-icu).

Example Failure: The issue was encountered on a system using the Thai (th-TH) culture, where the `RollingFileAppender` failed to create a backup due to an `ArgumentOutOfRangeException`. Below is the relevant error log:

```
LoggingService.Config: Trying to write default (embedded) file to standard path C:\ProgramData\Swing Catalyst\settings\log4net.xml
log4net:ERROR Could not create Appender [RollingFileAppender] of type [log4net.Appender.RollingFileAppender]. Reported error follows.
System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string. (Parameter 'startIndex')
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex)
   at log4net.Appender.RollingFileAppender.GetBackUpIndex(String curFileName)
   at log4net.Appender.RollingFileAppender.InitializeFromOneFile(String baseFile, String curFileName)
   at log4net.Appender.RollingFileAppender.InitializeRollBackups(String baseFile, ArrayList arrayFiles)
   at log4net.Appender.RollingFileAppender.DetermineCurSizeRollBackups()
   at log4net.Appender.RollingFileAppender.ExistingInit()
   at log4net.Appender.RollingFileAppender.ActivateOptions()
   at log4net.Repository.Hierarchy.XmlHierarchyConfigurator.ParseAppender(XmlElement appenderElement)
log4net:ERROR Appender named [RollingFileAppender] not found.
```

Suggested Fix: To avoid culture-specific errors, culture-agnostic string comparisons should be used throughout log4net. For example:

```csharp
someString.IndexOf("someStringValue", StringComparison.Ordinal)
```

Additional Resources:

* Performing Culture-Insensitive String Comparisons: [Culture-Insensitive String Comparisons](https://learn.microsoft.com/en-us/dotnet/core/extensions/performing-culture-insensitive-string-comparisons)

Related GitHub Issues for context:

* [Issue #59120](https://github.com/dotnet/runtime/issues/59120)
* [Issue #75616](https://github.com/dotnet/runtime/issues/75616)
* [Issue #76137](https://github.com/dotnet/runtime/issues/76137)

**Comment by FreeAndNil** [comment]
Hi @stianeklund,
Jira is no longer actively used (only for looking up old tickets).
Can you please verify, whether the issue still persists with 3.0.0?
We made a lot of changes (in RollingFileAppender) and switched the StringComparison to Ordinal everywhere in the class.

**Comment by stianeklund** [comment]
@FreeAndNil Good to know w/r/t Jira. Glad to hear this has likely been fixed in in 3.0. If string comparison uses ordinal everywhere in the `RollingFileAppender` it is indeed most likely fixed and you can consider closing this. I can reopen the issue if we somehow run into this again with v3. Testing this is a bit of a pain..
